components/waypoint-movement.js

```diff
@@ -0,0 +1,239 @@
+ THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+WL.registerComponent('waypoint-movement', {
+    /** This object is the container of the waypoints. It should contain children
+    with an alphebetically ascending naming (e.g. A, B, C, D, E, F) */
+    pathObject: {type: WL.Type.Object},
+    /** Movement speed of the object */
+    speed: {type: WL.Type.Float, default: 1.0},
+    /** Distance in normal space [0, 0.5) after which the objects starts moving on a curve */
+    curveDistance: {type: WL.Type.Float, default: 0.1},
+}, {
+    init: function() {
+      this.children = this.pathObject.children.sort(function(a, b) { return a.name.localeCompare(b.name) });
+      this.positions = new Array(this.children.length);
+      for (var i = 0; i < this.children.length; i++) {
+        this.positions[i] = new Float32Array(3);
+        this.children[i].getTranslationWorld(this.positions[i]);
+      }
+      this.currentPositionIndex = 0;
+      this.currentPosition = new Float32Array(3);
+      this.lookPosition = new Float32Array(3);
+      this.fromPosition = this.positions[0];
+      this.toPosition = this.positions[1];
+      glMatrix.vec3.sub(this.currentPosition, this.toPosition, this.fromPosition);
+      this.fromTolength = glMatrix.vec3.length(this.currentPosition);
+      this.currentLength = 0;
+      this.up = [0, 1, 0];
+      this.quat = new Float32Array(4);
+      lookAt(this.quat, this.toPosition, this.fromPosition, this.up);
+      this.object.resetTranslationRotation();
+      this.object.rotate(this.quat);
+      this.object.setTranslationWorld(this.fromPosition);
+
+      this.p0 = new Float32Array(3);
+      this.p1 = new Float32Array(3);
+      this.p2 = new Float32Array(3);
+      this.currentCurveIndex = 0;
+      this.curveDistance = Math.max(0.0, Math.min(0.49999, this.curveDistance));
+      this.bezFactorMultiplicator = 1 / (2*this.curveDistance);
+      this._setCurvePoints();
+
+      this.onFinalWaypointReachedCallbacks = [];
+    },
+    /**
+     * Adds a function to a callback array which, then calls the function after the final waypoint has been reached.
+     * @param  {[type]} f Function to be added.
+     */
+    addOnFinalWaypointReachedCallback: function(f) {
+      this.onFinalWaypointReachedCallbacks.push(f);
+    },
+    /**
+     * Removes a function from the callback array for reaching the final waypoint..
+     * @param  {[type]} f Function to be removed.
+     */
+    removeOnFinalWaypointReachedCallback: function(f) {
+      const index = array.indexOf(f);
+      if (index > -1) {
+        this.onFinalWaypointReachedCallbacks.splice(index, 1);
+      }
+    },
+    /**
+     * Sets the points needed for the next curve
+     */
+     _setCurvePoints: function(){
```

Comment:
There's one space too much indentation here 🤔 I'll set up a prettier config soon and reformat, so no need to care about that for now.